lib/stripe/stripe_client.rb

```diff
@@ -229,19 +229,36 @@ def format_app_info(info)
       str
     end
 
-    def handle_api_error(http_resp)
+    def handle_error_response(http_resp)
       begin
         resp = StripeResponse.from_faraday_hash(http_resp)
         error = resp.data[:error]
 
-        unless error && error.is_a?(Hash)
+        unless error
           raise StripeError.new("Indeterminate error")
         end
 
       rescue JSON::ParserError, StripeError
         raise general_api_error(http_resp[:status], http_resp[:body])
       end
 
+      if error.is_a?(Hash)
+        error = specific_api_error(resp, error)
+      elsif error.is_a?(String)
+        error = oauth_error(resp, error)
+      else
+        error = StripeError.new(
+          "Indeterminate error",
+          http_status: resp.http_status, http_body: resp.http_body,
+          json_body: resp.data, http_headers: resp.http_headers
+        )
+      end
```

Comment:
I completely agree with the sentiment and would rather that all our endpoints (whether API or OAuth) returned unambiguous errors. However, your proposal wouldn't work because `error_description` is optional :/

E.g.:

```
$ curl https://connect.stripe.com/oauth/token \
  -H "Authorization: Bearer $STRIPE_SECRET_KEY" \
  -d grant_type=foo
```

returns:
```
{
  "error": "unsupported_grant_type"
}
```

So AFAICT the only way to distinguish between API and OAuth errors by looking at the JSON response only is to check whether `error` is a hash or a string. The `StripeOAuthClient` class sort of solved this issue because it provided context: if the request was sent by `StripeOAuthClient` then any error that is returned is an OAuth error by default.

Without reintroducing the `StripeOAuthClient` class, we could still have some sort of request context that we could look at when handling errors. Do you think this would be preferable to just guessing based on `error`'s type?

Alternatively, we could fix the OAuth endpoints to always return an `error_description`. I didn't dig too deep but I think that the only error that lacks a description is `unsupported_grant_type` -- it shouldn't be too hard to add a static description everywhere.